### PR TITLE
Use CMake FetchContent to support both stable and HEAD ghidra

### LINKED ISSUE #14: Support both stable and latest (pinned commit) Ghidra
It would be nice to support not only the latest stable version of Ghidra but also secondary support for building Ghidra from the latest-ish commit on their `master` branch. This could be implemented through a CMake cache variable to set a `stable` or `HEAD` _Ghidra release type_ (i.e. `sleigh_GHIDRA_RELEASE_TYPE`).

While this would add a few conditional constructs in the CMake and potentially make it harder to read, I think it would be worth it to provide support for these two Ghidra release types. The conditionals would check the value of `sleigh_GHIDRA_RELEASE_TYPE` and do the right thing, with `HEAD` value as the moving target.

I can see this being implemented using CMake's [`FetchContent`](https://cmake.org/cmake/help/latest/module/FetchContent.html) module. The `FetchContent` module provides a nice way to pull the Ghidra repo, check out a commit, apply optional patches, and even use your own local copy of a Ghidra source repo for full control over what you are building. This allows for more flexible development of patches for upstream contribution. You can use your own Ghidra source directory during CMake configuration (with a clean build directory) by setting both 
```
-Dsleigh_GHIDRA_RELEASE_TYPE=HEAD -DFETCHCONTENT_SOURCE_DIR_GHIDRASOURCE=<path to your>/ghidra
```

I've taken a stab at implementing the source code fetching in a `cmake/setup-ghidra-source.cmake` file, which can be called in the top-level `CMakeLists.txt`.

I am happy to hook this all up with what exists right now, if you think this is a good approach

-------------------

`CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 3.21)

# Sets "library_root" variable for decompiler cpp root directory
# Use "ghidrasource_SOURCE_DIR" variable for Ghidra repo root
include(cmake/setup-ghidra-source.cmake)

project(
    sleigh
    VERSION ${sleigh_GHIDRA_VERSION}
    DESCRIPTION "Unofficial CMake build for Ghidra SLEIGH Decompiler"
    HOMEPAGE_URL "https://github.com/lifting-bits/sleigh"
    LANGUAGES CXX
)
...
```
---------------------
`cmake/setup-ghidra-source.cmake`
```cmake
# ---- Setup Ghidra Source code ----

# Set up Ghidra repo human-readable version settings
set(sleigh_GHIDRA_RELEASE_TYPE "stable" CACHE
  STRING "Ghidra release type to use. Make sure sleigh_GHIDRA_COMMIT is the correct corresponding commit. 'HEAD' is used for active development purposes."
)
# This is just helper for CMake UIs. CMake does not enforce that the value matches one of those listed.
set_property(CACHE sleigh_GHIDRA_RELEASE_TYPE PROPERTY STRINGS "stable" "HEAD")

# **** Setup pinned git info ****

# Ghidra pinned stable version commit
set(ghidra_version "10.0.4")
set(ghidra_git_tag "5b07797cb859d8801ca3a1e08bda1321ca3ce002")
# pinned stable patches list
set(ghidra_patches "")

# Ghidra pinned commits used for pinning last known working HEAD commit
if("${sleigh_GHIDRA_RELEASE_TYPE}" STREQUAL HEAD)
  # TODO: Try to remember to look at Ghidra/application.properties
  # TODO: CMake only likes numeric characters in the version string....
  set(ghidra_version "10.1")
  set(ghidra_git_tag "55b8fcf7d4aeaca37ffb5c947340915d69c84224")
  set(ghidra_patches
    PATCH_COMMAND git am --ignore-space-change --ignore-whitespace --no-gpg-sign
      "${CMAKE_CURRENT_SOURCE_DIR}/patches/HEAD/0001-Fix-arg-parsing-in-sleigh-C-test-runner.patch"
  )
endif()

# For use in the CMake `project` command to set version
set(sleigh_GHIDRA_VERSION "${ghidra_version}" CACHE
  STRING "Numeric Ghidra version corresponding to sleigh_GHIDRA_RELEASE_TYPE and sleigh_GHIDRA_COMMIT. This is used during packaging"
)

set(sleigh_GHIDRA_COMMIT "${ghidra_git_tag}" CACHE
  STRING "Ghidra repo commit to use/checkout. Ensure this correct with respect to sleigh_GHIDRA_VERSION and sleigh_GHIDRA_RELEASE_TYPE."
)
string(SUBSTRING "${ghidra_git_tag}" 0 7 ghidra_short_commit)

message(STATUS "Using Ghidra version ${sleigh_GHIDRA_VERSION} at commit ${ghidra_short_commit}")

include(FetchContent)
# Write out source directory with identifiable version info
FetchContent_Declare(GhidraSource
    GIT_REPOSITORY https://github.com/NationalSecurityAgency/ghidra
    GIT_TAG "${sleigh_GHIDRA_COMMIT}"
    GIT_PROGRESS TRUE
    SOURCE_DIR "${CMAKE_SOURCE_DIR}/source_${sleigh_GHIDRA_VERSION}_${ghidra_short_commit}"
    ${ghidra_patches}
)
FetchContent_MakeAvailable(GhidraSource)

message(STATUS "Ghidra source located at '${ghidrasource_SOURCE_DIR}'")

# Sanity check on Ghidra source code checkout
set(library_root "${ghidrasource_SOURCE_DIR}/Ghidra/Features/Decompiler/src/decompile/cpp")
if(NOT EXISTS "${library_root}/sleigh.hh")
  message(FATAL_ERROR "The Ghidra source directory has not been initialized correctly. Could not find '${library_root}'")
endif()
```

### PR DESCRIPTION
**Pull request by ekilmer**
closes #14 

TODO:

- [x] More testing to make sure this method plays nice with different workflows of people (especially the `FETCHCONTENT_BASE_DIR`)
- [x] Updating the README to make sure people know what's going on and/or how to use their own Ghidra source checkout.

Would appreciate people testing it out and giving comments!

Another thing that would be nice to have eventually is weekly testing of new `HEAD` commits from Ghidra main repo. To be fully automated, I think we would have to do something similar to dependabot but update the `src/setup-ghidra-source.cmake` line with
```cmake
set(ghidra_git_tag "<head commit>")
```

Would also be nice to auto-generate the slaspec file listings.

### CODE REVIEW

File: src/setup-ghidra-source.cmake
Comment (tetsuo-cpp) [inline src/setup-ghidra-source.cmake:18]: Do we have any other builds that work like this? I'm just wondering how we usually manage `HEAD` in a setup like this. Maybe I should periodically bump it to the tip of `master` and make sure it builds ok.

**Comment by tetsuo-cpp** [comment]
Thanks, this is awesome. I had a problem switching to `HEAD` after I'd built `stable`.
```
tetsuo@Alexs-MacBook-Pro sleigh % cmake -B build -S . -DSLEIGH_ENABLE_INSTALL=ON -DSLEIGH_GHIDRA_RELEASE_TYPE=HEAD
-- Using Ghidra version 10.0.4 at commit 55b8fcf
-- Populating ghidrasource
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/tetsuo/Code/sleigh/build/_deps/ghidrasource-subbuild
[0/7] Performing update step for 'ghidrasource-populate'
[2/7] Performing patch step for 'ghidrasource-populate'
FAILED: ghidrasource-populate-prefix/src/ghidrasource-populate-stamp/ghidrasource-populate-patch /Users/tetsuo/Code/sleigh/build/_deps/ghidrasource-subbuild/ghidrasource-populate-prefix/src/ghidrasource-populate-stamp/ghidrasource-populate-patch
cd /Users/tetsuo/Code/sleigh/build/_deps/ghidrasource-src && git am --ignore-space-change --ignore-whitespace --no-gpg-sign /Users/tetsuo/Code/sleigh/patches/HEAD/0001-Fix-arg-parsing-in-sleigh-C-test-runner.patch && /opt/homebrew/Cellar/cmake/3.21.3/bin/cmake -E touch /Users/tetsuo/Code/sleigh/build/_deps/ghidrasource-subbuild/ghidrasource-populate-prefix/src/ghidrasource-populate-stamp/ghidrasource-populate-patch
fatal: previous rebase directory .git/rebase-apply still exists but mbox given.
ninja: build stopped: subcommand failed.
CMake Error at /opt/homebrew/Cellar/cmake/3.21.3/share/cmake/Modules/FetchContent.cmake:1017 (message):
  Build step for ghidrasource failed: 1
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.21.3/share/cmake/Modules/FetchContent.cmake:1146:EVAL:2 (__FetchContent_directPopulate)
  /opt/homebrew/Cellar/cmake/3.21.3/share/cmake/Modules/FetchContent.cmake:1146 (cmake_language)
  /opt/homebrew/Cellar/cmake/3.21.3/share/cmake/Modules/FetchContent.cmake:1189 (FetchContent_Populate)
  src/setup-ghidra-source.cmake:54 (FetchContent_MakeAvailable)
  CMakeLists.txt:13 (include)


-- Configuring incomplete, errors occurred!
See also "/Users/tetsuo/Code/sleigh/build/CMakeFiles/CMakeOutput.log".
See also "/Users/tetsuo/Code/sleigh/build/CMakeFiles/CMakeError.log".
```
It seems to be trying to apply the patch even though the Ghidra source hasn't been checked out to the `HEAD` commit, and therefore failing. I had a quick look at the [FetchContent docs](https://cmake.org/cmake/help/latest/module/FetchContent.html) but didn't see anything obvious.

**Comment by ekilmer** [comment]
> I had a problem switching to HEAD after I'd built stable.

Ahh, yeah this is a known issue due to the way I implemented the logic. Basically, I used more cache variables for specifying the commit SHA and numeric Ghidra version number, which would also need to be changed accordingly upon re-configuration from a `stable` release to a `HEAD` release. If you try the following sequence, it should work.

```bash
$ cmake -B build -S . -DSLEIGH_ENABLE_INSTALL=ON
...
$ cmake --build build -j
...
$ cmake -B build -S . -DSLEIGH_ENABLE_INSTALL=ON \
    -DSLEIGH_GHIDRA_RELEASE_TYPE=HEAD \
    -DSLEIGH_GHIDRA_VERSION=10.1 \
    -DSLEIGH_GHIDRA_COMMIT=55b8fcf7d4aeaca37ffb5c947340915d69c84224
...
$ cmake --build build -j
...
```

I think the workflow might be a matter of preference, but I'm happy to discuss. I see two options for resolving this:

1. Tell people in the README that they should use separate build directories for stable and HEAD builds; or if they want to reconfigure the same build directory, then they must specify the other details as they appear in the setup script.
2. Remove the other cache variables leaving only the `SLEIGH_GHIDRA_RELEASE_TYPE` cache variable. This reduces the ability to customize without changing the source, but it also makes our lives easier by not supporting a workflow that involves some somewhat hidden details, as you've found out.

I'm actually leaning towards option 2 because it reduces our pain of dependent cache variables and makes the configuration a bit more streamlined. If an end user wants to use their own arbitrary checkouts, then they should use the method supported by FetchContent to point to their own source directory. That method also makes it more obvious that what they're doing likely isn't officially supported. Furthermore, updating to a different `HEAD` version requires a change to the source code, which could be better from a visibility standpoint when testing upstream changes.

Thoughts?

File: README.md
Comment (ekilmer) [reply README.md:28]: Yeah, so my thought process was that the "Required" dependencies are those that are _required_ during a minimal CMake invocation:

```bash
cmake -B build -S .
```

which doesn't turn on the installation options by default, thus only requiring .

If we want installation turned on by default, then we can move these up.

I can add notes about why some of these are needed as well. Thoughts?

Comment (ekilmer) [reply README.md:32]: Added in 23a021f8a7a7dd619e78a0a59535af7ee1586c1f

File: src/setup-ghidra-source.cmake
Comment (ekilmer) [reply src/setup-ghidra-source.cmake:18]: Honestly, I'm not aware of any other projects doing something like this.

I think some sort of bot or GitHub Action would be the best way to keep track of this. Sort of like Dependabot but we would have to script it ourselves. I see the workflow like:

1. Create a weekly GitHub Action cron job
2. Grab the latest commit from default development Ghidra branch (`master`)
3. Check that it isn't the same commit as what we have currently
4. Write new commit SHA to the `./src/setup-ghidra-source.cmake` HEAD commit variable
5. Commit and push as new PR
6. Let CI run
7. Wait for manual approval

It would also be nice to automatically update the listing of `slaspec` files...

Comment (tetsuo-cpp) [reply src/setup-ghidra-source.cmake:18]: Thanks. That sounds like a good idea.

**Review by tetsuo-cpp — APPROVED** [review]
LGTM! Thanks, this works great. Feel free to merge whenever you're ready.

**Pull request merged**
